File: src/services/settingsService.ts
/** Hotel information, room-independent preferences and the optional app PIN. */

import { AppError } from '../utils/errors';
import { validatePin } from '../utils/validation';
import type { SettingsRepository } from '../database/repositories/settingsRepository';
import type { HotelSettings } from '../types';

const DEFAULTS: HotelSettings = {
  hotel_name: 'My Hotel',
  hotel_address: '',
  hotel_phone: '',
  pin_hash: '',
  last_backup_at: '',
};

/** Domain separation so the stored digest is useless elsewhere. */
const PIN_SALT = 'hotel-customer-manager/pin/v1:';

function fallbackHash(input: string): string {
  // Only used if Web Crypto is unavailable (very old WebView). Still local-only.
  let h1 = 0x811c9dc5;
  let h2 = 0x01000193;
  for (let i = 0; i < input.length; i += 1) {
    h1 = (h1 ^ input.charCodeAt(i)) >>> 0;
    h1 = Math.imul(h1, 16777619) >>> 0;
    h2 = (h2 + Math.imul(input.charCodeAt(i) + i, 2654435761)) >>> 0;
  }
  return `fnv1a$${h1.toString(16).padStart(8, '0')}${h2.toString(16).padStart(8, '0')}`;
}

export async function hashPin(pin: string): Promise<string> {
  const value = PIN_SALT + pin;
  const subtle = globalThis.crypto?.subtle;
  if (!subtle) return fallbackHash(value);
  const digest = await subtle.digest('SHA-256', new TextEncoder().encode(value));
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, '0')).join('');
}

export class SettingsService {
  constructor(private readonly repo: SettingsRepository) {}

  async get(): Promise<HotelSettings> {
    const stored = await this.repo.getAll();
    return {
      hotel_name: stored.hotel_name || DEFAULTS.hotel_name,
      hotel_address: stored.hotel_address ?? DEFAULTS.hotel_address,
      hotel_phone: stored.hotel_phone ?? DEFAULTS.hotel_phone,
      pin_hash: stored.pin_hash ?? '',
      last_backup_at: stored.last_backup_at ?? '',
    };
  }

  async saveHotelInfo(info: Pick<HotelSettings, 'hotel_name' | 'hotel_address' | 'hotel_phone'>): Promise<void> {
    const name = info.hotel_name.trim();
    if (!name) throw new AppError('VALIDATION', 'Hotel name is required.');
    if (name.length > 80) throw new AppError('VALIDATION', 'Hotel name is too long.');
    await this.repo.setMany({
      hotel_name: name,
      hotel_address: info.hotel_address.trim(),
      hotel_phone: info.hotel_phone.trim(),
    });
  }

  async isPinEnabled(): Promise<boolean> {
    return Boolean(await this.repo.get('pin_hash'));
  }

  async setPin(pin: string): Promise<void> {
    const error = validatePin(pin);
    if (error) throw new AppError('VALIDATION', error);
    await this.repo.set('pin_hash', await hashPin(pin));
  }

  async verifyPin(pin: string): Promise<boolean> {
    const stored = await this.repo.get('pin_hash');
    if (!stored) return true; // no PIN configured
    return stored === (await hashPin(pin));
  }

  /** Removing the PIN requires the current one. */
  async clearPin(currentPin: string): Promise<void> {
    if (!(await this.verifyPin(currentPin))) {
      throw new AppError('VALIDATION', 'That PIN is incorrect.');
    }
    await this.repo.set('pin_hash', '');
  }

  async lastBackupAt(): Promise<string> {
    return (await this.repo.get('last_backup_at')) ?? '';
  }
}

File: lib/services/settings_service.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

import '../database/repositories/settings_repository.dart';
import '../models/models.dart';
import '../utils/app_error.dart';
import '../utils/message_template.dart';
import '../utils/validation.dart';

const _pinSalt = 'hotel-customer-manager/pin/v1:';

const _defaults = HotelSettings(
  hotelName: 'My Hotel',
  hotelAddress: '',
  hotelPhone: '',
  pinHash: '',
  lastBackupAt: '',
  whatsappEnabled: true,
  whatsappCountryCode: '91',
  whatsappWelcomeTemplate: defaultWelcomeTemplate,
);

/// Longest welcome message we'll store. `wa.me` prefill is a URL query
/// parameter, so a runaway template would get truncated by the OS anyway.
const _maxTemplateLength = 1000;

final _countryCodePattern = RegExp(r'^\d{1,4}$');

String _hashPin(String pin) => sha256.convert(utf8.encode('$_pinSalt$pin')).toString();

/// Hotel identity, and the optional app PIN.
///
/// The PIN itself is **never stored** — only a salted SHA-256 digest. An
/// empty digest means "no PIN configured", in which case [verifyPin] always
/// succeeds (there's nothing to check against).
class SettingsService {
  final SettingsRepository _repo;
  SettingsService(this._repo);

  Future<HotelSettings> get() async {
    final stored = await _repo.getAll();
    return HotelSettings(
      hotelName: (stored['hotel_name']?.isNotEmpty ?? false) ? stored['hotel_name']! : _defaults.hotelName,
      hotelAddress: stored['hotel_address'] ?? _defaults.hotelAddress,
      hotelPhone: stored['hotel_phone'] ?? _defaults.hotelPhone,
      pinHash: stored['pin_hash'] ?? _defaults.pinHash,
      lastBackupAt: stored['last_backup_at'] ?? _defaults.lastBackupAt,
      // These three keys are absent on installs created before the WhatsApp
      // feature shipped, so they fall back rather than needing a migration.
      whatsappEnabled: (stored['whatsapp_enabled'] ?? '1') == '1',
      whatsappCountryCode: (stored['whatsapp_country_code']?.isNotEmpty ?? false)
          ? stored['whatsapp_country_code']!
          : _defaults.whatsappCountryCode,
      whatsappWelcomeTemplate: (stored['whatsapp_welcome_template']?.isNotEmpty ?? false)
          ? stored['whatsapp_welcome_template']!
          : _defaults.whatsappWelcomeTemplate,
    );
  }

  Future<void> saveHotelInfo({
    required String hotelName,
    required String hotelAddress,
    required String hotelPhone,
  }) async {
    final name = hotelName.trim();
    if (name.isEmpty || name.length > 80) {
      throw const AppError(AppErrorCode.validation, 'Hotel name is required (up to 80 characters).');
    }
    await _repo.setMany({
      'hotel_name': name,
      'hotel_address': hotelAddress.trim(),
      'hotel_phone': hotelPhone.trim(),
    });
  }

  Future<void> saveWhatsAppSettings({
    required bool enabled,
    required String countryCode,
    required String template,
  }) async {
    final code = countryCode.trim().replaceFirst(RegExp(r'^\+'), '');
    if (!_countryCodePattern.hasMatch(code)) {
      throw const AppError(AppErrorCode.validation, 'Country code must be 1-4 digits.');
    }
    final body = template.trim();
    if (body.isEmpty || body.length > _maxTemplateLength) {
      throw const AppError(
        AppErrorCode.validation,
        'Welcome message is required (up to $_maxTemplateLength characters).',
      );
    }
    await _repo.setMany({
      'whatsapp_enabled': enabled ? '1' : '0',
      'whatsapp_country_code': code,
      'whatsapp_welcome_template': body,
    });
  }

  Future<bool> isPinEnabled() async => ((await _repo.get('pin_hash')) ?? '').isNotEmpty;

  Future<void> setPin(String pin) async {
    final error = validatePin(pin);
    if (error != null) throw AppError(AppErrorCode.validation, error);
    await _repo.set('pin_hash', _hashPin(pin));
  }

  Future<bool> verifyPin(String pin) async {
    final stored = await _repo.get('pin_hash') ?? '';
    if (stored.isEmpty) return true;
    return stored == _hashPin(pin);
  }

  Future<void> clearPin(String currentPin) async {
    if (!await verifyPin(currentPin)) {
      throw const AppError(AppErrorCode.validation, 'That PIN is incorrect.');
    }
    await _repo.set('pin_hash', '');
  }

  Future<String> lastBackupAt() async => (await _repo.get('last_backup_at')) ?? '';
}
